### PR TITLE
Retransmitting ACK frames can inflate RTT

### LINKED ISSUE #2748: Retransmitting ACKs may cause misleading RTT samples
Even though a sender is encouraged to assemble frames containing up-to-date information every time it sends a packet, it is not forbidden to retransmit copies of frames from lost packets. A receiver MUST accept packets containing an outdated frame, such as a MAX_DATA frame carrying a smaller maximum data than one found in an older packet.
https://github.com/quicwg/base-drafts/pull/2312/files

If a packet containing an ACK is lost and the payload is retransmitted with a new packet number AND no ACKs are received prior to the retransmitted ACK frame arriving, the peer will calculate and RTT sample that is 2 RTTs + loss recovery delay.

As such, I'd suggest we add a warning that retransmitting ACKs can cause RTT inflation and is heavily discouraged.

There may be other issues with retransmitting ACKs, but this is the first that comes to mind.

### PR DESCRIPTION
**Pull request by ianswett**
Fixes #2748

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -291,6 +291,13 @@ are received within an RTT.  As suggested in {{?RFC6298}}, doing so might result
 in inadequate history in smoothed_rtt and rttvar.  Ensuring that RTT estimates
 retain sufficient history is an open research question.
 
+### Retransmitted ACK frames
+
+Section 13.3 of {{QUIC-TRANSPORT}} recommends sending the most recent set of
+acknowledgements in ACK frames, but if ACK frames are retransmitted, an inflated
+RTT can be measured.  Some incorrect RTT samples are prevented by only measuring
+RTT when the largest acked increases.
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: Type the entire word:

```suggestion
RTT when the largest acknowledged increases.
```

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -291,6 +291,13 @@ are received within an RTT.  As suggested in {{?RFC6298}}, doing so might result
 in inadequate history in smoothed_rtt and rttvar.  Ensuring that RTT estimates
 retain sufficient history is an open research question.
 
+### Retransmitted ACK frames
+
+Section 13.3 of {{QUIC-TRANSPORT}} recommends sending the most recent set of
+acknowledgements in ACK frames, but if ACK frames are retransmitted, an inflated
```
Comment (martinthomson) [inline draft-ietf-quic-recovery.md]:
```suggestion
acknowledgements in ACK frames, but if ACK frames are retransmitted without adjusting the Ack Delay, an inflated
```

**Comment by janaiyengar** [comment]
I agree with @martinthomson. Perhaps it's better to add some text where we discourage retransmitting of frames (at the end of [Section 13.3](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.13.3)) instead of in recovery.

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (ianswett) [reply draft-ietf-quic-recovery.md]: I originally wrote that, but it's likely to be inflated anyway, particularly if max_ack_delay is < RTT.

**Comment by ianswett** [comment]
I created #3049 as an alternative.  Or we can close this issue and move on.

**Comment by mikkelfj** [comment]
I agree that ACKs should be sent up to date, but I also think it makes sense the the receiver to protect itself in this regard. Some implementations will mindlessly repeat frames and it is not enough to cause a protocol violation.

**Pull request closed without merging**